plutus-emulator/src/Wallet/Emulator/Types.hs

```diff
@@ -280,16 +284,21 @@ instance WalletAPI MockWallet where
         privK <- use ownPrivateKey
         pure (Crypto.sign (BSL.toStrict bs) privK)
 
-    createPaymentWithChange vl = do
+    updatePaymentWithChange vl (oldIns, changeTx) = do
         ws <- get
-        let fnds   = ws ^. ownFunds
+        let usedFnds = Set.map txInRef oldIns
+            oldChange = maybe (Ada.lovelaceValueOf 0) txOutValue changeTx
+            fnds   = Map.withoutKeys (ws ^. ownFunds) usedFnds
             privK  = ws ^. ownPrivateKey
-            pubK   =  toPublicKey privK
-        (spend, change) <- selectCoin (second txOutValue <$> Map.toList fnds) vl
-        let
-            txOutput = if Value.isZero change then Nothing else Just (pubKeyTxOut change pubK)
-            ins = Set.fromList (flip pubKeyTxIn pubK . fst <$> spend)
-        pure (ins, txOutput)
+            pubK   = toPublicKey privK
+        if vl `Value.leq` oldChange
+        then pure (oldIns, mkChangeOutput pubK $ oldChange PlutusTx.- vl)
+        else do
+          (spend, change) <- selectCoin (second txOutValue <$> Map.toList fnds)
+                                        (vl PlutusTx.- oldChange)
+          let
+              ins = Set.fromList (flip pubKeyTxIn pubK . fst <$> spend)
```

Comment:
The only non-saturated uses of `pubKeyTxIn` I can see call `flip` on it. Therefore I think we should just change the order of the arguments.

Reply:
Note that this breaks the book. I will also update it.

Reply:
Yes, just update the uses.